File: node/engines/bing.js
const { getBrowser } = require("../_browser");

async function fetchBing(query, n = 10) {
  const browser = await getBrowser();
  const page = await browser.newPage();
  try {
    await page.goto(
      `https://www.bing.com/images/search?q=${encodeURIComponent(query)}&form=HDRSC2`,
      { waitUntil: "domcontentloaded" }
    );
    return await page.$$eval(
      "a.iusc",
      (links, maxN) => {
        const urls = [];
        const seen = new Set();
        for (const link of links) {
          try {
            const m = JSON.parse(link.getAttribute("m") || "{}");
            if (m.murl && !seen.has(m.murl)) {
              seen.add(m.murl);
              urls.push(m.murl);
              if (urls.length >= maxN) break;
            }
          } catch {}
        }
        return urls;
      },
      n
    );
  } finally {
    await page.close();
  }
}

module.exports = { fetchBing };

File: imgapi/engines/bing.py
import json
from urllib.parse import quote_plus
from .._fetcher import get_fetcher


def fetch_bing(query: str, n: int = 10) -> list[str]:
    page = get_fetcher().fetch(
        f"https://www.bing.com/images/search?q={quote_plus(query)}&form=HDRSC2",
        headless=True,
    )
    urls: list[str] = []
    for link in page.css("a.iusc"):
        try:
            m = json.loads(link.attrib.get("m", "{}"))
            url = m.get("murl", "")
            if url and url not in urls:
                urls.append(url)
                if len(urls) >= n:
                    break
        except (json.JSONDecodeError, AttributeError):
            continue
    return urls
